syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt

```diff
@@ -14,145 +15,175 @@
+            // delete all temp blocks now
+            tempRepository.deleteTempData(blockTempIdByHash.values.toList())
+
+            throw ex
         }
     }
 
-    abstract inner class FileDownloadObserver : Closeable {
+    private suspend fun pullBlock(fileBlocks: FileBlocks, block: BlockInfo, timeoutInMillis: Long): ByteArray {
+        logger.debug("sent request for block, hash = {}", block.hash)
+
+        val response =
+                withTimeout(timeoutInMillis) {
+                    try {
+                        doRequest(
+                                Request.newBuilder()
+                                        .setFolder(fileBlocks.folder)
+                                        .setName(fileBlocks.path)
+                                        .setOffset(block.offset)
+                                        .setSize(block.size)
+                                        .setHash(ByteString.copyFrom(Hex.decode(block.hash)))
+                        )
+                    } catch (ex: TimeoutCancellationException) {
+                        // It seems like the TimeoutCancellationException
+                        // is handled differently so that the timeout is ignored.
+                        // Due to that, it's converted to an IOException.
+
+                        throw IOException("timeout during requesting block")
+                    }
+                }
 
-        abstract fun progress(): Double
+        NetworkUtils.assertProtocol(response.code == ErrorCode.NO_ERROR) {
+            "received error response, code = ${response.code}"
+        }
 
-        abstract fun progressMessage(): String
+        val data = response.data.toByteArray()
+        val hash = Hex.toHexString(MessageDigest.getInstance("SHA-256").digest(data))
 
-        abstract fun isCompleted(): Boolean
+        if (hash != block.hash) {
+            throw IllegalStateException("expected block with hash ${block.hash}, but got block with hash $hash")
+        }
 
-        abstract fun inputStream(): InputStream
+        return data
+    }
 
-        abstract fun checkError()
+    private suspend fun doRequest(request: Request.Builder): BlockExchangeProtos.Response {
+        return suspendCancellableCoroutine {
+            continuation ->
 
-        @Throws(InterruptedException::class)
-        abstract fun waitForProgressUpdate(): Double
+            val requestId = responseHandler.registerListener {
+                response ->
 
-        @Throws(InterruptedException::class)
-        fun waitForComplete(): FileDownloadObserver {
-            while (!isCompleted()) {
-                waitForProgressUpdate()
+                continuation.resume(response)
             }
-            return this
-        }
-
-        abstract override fun close()
 
+            connectionHandler.sendMessage(
+                    request
+                            .setId(requestId)
+                            .build()
+            )
```

Comment:
Feels like wonky formatting again.

Reply:
Same again. It's already the result of the automatic reformat.